Freshli/Languages/CSharp/NuGetRepository.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Freshli.Exceptions;
+using NuGet.Common;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace Freshli.Languages.CSharp {
+  public class NuGetRepository : IPackageRepository {
+
+    private IDictionary<string, IList<IVersionInfo>> _packages =
+      new Dictionary<string, IList<IVersionInfo>>();
+
+    private IEnumerable<IVersionInfo> GetReleaseHistory(
+      string name,
+      bool includePreReleaseVersions) {
+      var versions = GetMetadata(name);
+
+      return versions.Select(v => new FreshliNuGetVersionInfo(v));
+    }
+
+    private IEnumerable<IPackageSearchMetadata> GetMetadata(string name) {
+      ILogger logger = NullLogger.Instance;
+      CancellationToken cancellationToken = CancellationToken.None;
+
+      SourceCacheContext cache = new SourceCacheContext();
+      SourceRepository repository = Repository.Factory.GetCoreV3(
+        "https://api.nuget.org/v3/index.json"
+      );
+      PackageMetadataResource resource =
+        repository.GetResourceAsync<PackageMetadataResource>().Result;
+
+      IEnumerable<IPackageSearchMetadata> packages = resource.GetMetadataAsync(
+          name,
+          includePrerelease: true,
+          includeUnlisted: false,
+          cache,
+          logger,
+          cancellationToken).Result;
+
+      return packages;
+    }
+
+    public IVersionInfo Latest(
+      string name,
+      DateTime asOf,
+      bool includePreReleases)
+    {
+      try {
+        return GetReleaseHistory(name, includePreReleases).
+          OrderByDescending(v => v.DatePublished).
+          First(v => asOf >= v.DatePublished);
+      } catch (Exception e) {
+        throw new LatestVersionNotFoundException(name, asOf, e);
+      }
+    }
+
+    public IVersionInfo VersionInfo(string name, string version) {
+      try {
+        return GetReleaseHistory(name, includePreReleaseVersions: true).
+          First(v => v.Version == version);
+      } catch (Exception e) {
+        throw new VersionNotFoundException(name, version, e);
+      }
+    }
+
+    public List<IVersionInfo> VersionsBetween(
+      string name,
```

Comment:
Going to address this in a later issue, since this is going to take some effort to remediate this issue: https://github.com/corgibytes/freshli/issues/236

Reply:
@dfar-io can you add an ignore tag and a comment that this fails CodeClimate but will be fixed in a later issue.  Assuming there is a way to tell CodeClimate to ignore errors.

Reply:
The closest I could find was ignoring the file as a whole - not ideal but it'll work for the time being, will just need to remember to turn off the ignore when fixing above issue.